compose/ui/ui-text/src/skikoMain/kotlin/androidx/compose/ui/text/platform/ParagraphLayouter.skiko.kt

```diff
@@ -61,49 +68,99 @@ class ParagraphLayouter(
-        maxLines: Int = builder.maxLines,
-        ellipsis: String = builder.ellipsis,
-        color: Color = builder.textStyle.color,
-        shadow: Shadow? = builder.textStyle.shadow,
-        textDecoration: TextDecoration? = builder.textStyle.textDecoration,
-    ): Paragraph {
+    fun setParagraphStyle(
+        maxLines: Int,
+        ellipsis: String
+    ) {
+        if (builder.maxLines != maxLines ||
+            builder.ellipsis != ellipsis
+        ) {
+            builder.maxLines = maxLines
+            builder.ellipsis = ellipsis
+            paragraphCache = null
+        }
+    }
+
+    fun setTextStyle(
+        color: Color,
+        shadow: Shadow?,
+        textDecoration: TextDecoration?
+    ) {
         val actualColor = color.takeOrElse { builder.textStyle.color }
-        if (
-            builder.maxLines != maxLines ||
-            builder.ellipsis != ellipsis ||
-            builder.textStyle.color != actualColor ||
+        if (builder.textStyle.color != actualColor ||
             builder.textStyle.shadow != shadow ||
             builder.textStyle.textDecoration != textDecoration
         ) {
-            this.width = width
-            builder.maxLines = maxLines
-            builder.ellipsis = ellipsis
             builder.textStyle = builder.textStyle.copy(
                 color = actualColor,
                 shadow = shadow,
                 textDecoration = textDecoration
             )
-            para = builder.build()
-            para.layout(width)
-        } else if (this.width != width) {
-            this.width = width
-            para.layout(width)
+            paragraphCache = null
         }
-        return para
     }
+
+    @ExperimentalTextApi
+    fun setTextStyle(
+        brush: Brush?,
+        brushSize: Size,
+        alpha: Float,
+        shadow: Shadow?,
+        textDecoration: TextDecoration?
+    ) {
+        val actualSize = builder.brushSize
+        if (builder.textStyle.brush != brush ||
+            actualSize == null ||
+            !actualSize.width.sameValueAs(brushSize.width) ||
+            !actualSize.height.sameValueAs(brushSize.height) ||
+            !builder.textStyle.alpha.sameValueAs(alpha) ||
+            builder.textStyle.shadow != shadow ||
+            builder.textStyle.textDecoration != textDecoration
+        ) {
+            builder.textStyle = builder.textStyle.copy(
+                brush = brush,
+                alpha = alpha,
+                shadow = shadow,
+                textDecoration = textDecoration
+            )
+            builder.brushSize = brushSize
+            paragraphCache = null
+        }
+    }
+
+    fun setDrawStyle(drawStyle: DrawStyle?) {
+        // TODO Implement applying DrawStyle
+    }
+
+    fun layoutParagraph(width: Float): Paragraph {
+        val paragraph = paragraphCache
+        return if (paragraph != null) {
+            if (!this.width.sameValueAs(width)) {
+                this.width = width
```

Comment:
I think here `sameValueAs` is redundant too.
width saves as is.

Is `this.width` modifies somewhere?

Reply:
The comparison wasn't introduced in this PR, but comparing floats with `==` was incorrect.

Reply:
Ok